File: miit_rule_checker/src/main/java/com/loper7/miit_rule_checker/MIITMethods.java
package com.loper7.miit_rule_checker;

import android.content.ContentResolver;
import android.location.LocationListener;
import android.provider.Settings;

import java.lang.reflect.Member;
import java.util.ArrayList;

/**
 * @Author loper7
 * @Date 2023/12/6 10:17
 * @Description
 **/
public class MIITMethods {

    public static ArrayList<Member> getDefaultMethods() {
        ArrayList<Member> members = new ArrayList<>();
        members.add(WifiInfo.getMacAddress);
        members.add(WifiInfo.getIpAddress);
        members.add(LocationManager.getLastKnownLocation);
        members.add(LocationManager.requestLocationUpdates);
        members.add(NetworkInterface.getHardwareAddress);
        members.add(ApplicationPackageManager.getInstalledPackages);
        members.add(ApplicationPackageManager.getInstalledApplications);
        members.add(ApplicationPackageManager.getPackageInfo);
        members.add(ApplicationPackageManager.getInstallerPackageName);
        members.add(Secure.getString);
        members.add(TelephonyManager.getDeviceId);
        members.add(TelephonyManager.getDeviceIdWithInt);
        members.add(TelephonyManager.getImei);
        members.add(TelephonyManager.getImeiWithInt);
        members.add(TelephonyManager.getSubscriberId);
        return members;
    }

    public static ArrayList<Member> getPackageManagerAll() {
        ArrayList<Member> members = new ArrayList<>();
        members.add(ApplicationPackageManager.getInstalledPackages);
        members.add(ApplicationPackageManager.getInstalledApplications);
        members.add(ApplicationPackageManager.getPackageInfo);
        members.add(ApplicationPackageManager.getInstallerPackageName);
        return members;
    }

    public static ArrayList<Member> getStringAll() {
        ArrayList<Member> members = new ArrayList<>();
        members.add(Secure.getString);
        members.add(System.getString);
        return members;
    }

    public static class WifiInfo {
        public static final Class<?> clazz = android.net.wifi.WifiInfo.class;

        public static Member getMacAddress = null;

        static {
            try {
                getMacAddress = clazz.getDeclaredMethod("getMacAddress");
            } catch (NoSuchMethodException ignored) {

            }
        }

        public static Member getIpAddress;

        static {
            getIpAddress = null;
            try {
                getIpAddress = clazz.getDeclaredMethod("getIpAddress");
            } catch (NoSuchMethodException ignored) {
            }
        }

    }

    public static class LocationManager {
        public static final Class<?> clazz = android.location.LocationManager.class;

        public static Member getLastKnownLocation = null;
        public static Member requestLocationUpdates = null;

        static {
            try {
                getLastKnownLocation = clazz.getDeclaredMethod("getLastKnownLocation", String.class);
                requestLocationUpdates = clazz.getDeclaredMethod("requestLocationUpdates", String.class,
                        long.class,
                        float.class,
                        LocationListener.class);
            } catch (NoSuchMethodException e) {

            }
        }
    }

    public static class NetworkInterface {
        public static final Class<?> clazz = java.net.NetworkInterface.class;

        public static Member getHardwareAddress = null;

        static {
            try {
                getHardwareAddress = clazz.getDeclaredMethod("getHardwareAddress");
            } catch (NoSuchMethodException e) {
            }
        }
    }

    public static class ApplicationPackageManager {
        public static Class<?> clazz = null;
        public static Member getInstallerPackageName = null;
        public static Member getInstalledApplications = null;
        public static Member getInstalledPackages = null;
        public static Member getPackageInfo = null;

        static {
            try {
                clazz = Class.forName("android.app.ApplicationPackageManager");
                getInstalledPackages = clazz.getDeclaredMethod("getInstalledPackages", int.class);
                getInstalledApplications = clazz.getDeclaredMethod("getInstalledApplications", int.class);
                getInstallerPackageName = clazz.getDeclaredMethod("getInstallerPackageName", String.class);
                getPackageInfo = clazz.getDeclaredMethod("getPackageInfo", String.class, int.class);
            } catch (Exception e) {
            }
        }
    }

    public static class Secure {
        public static final Class<?> clazz = android.provider.Settings.Secure.class;

        public static Member getString = null;

        static {
            try {
                getString = clazz.getDeclaredMethod("getString", ContentResolver.class, String.class);
            } catch (NoSuchMethodException e) {
            }
        }
    }

    public static class System {
        public static final Class<?> clazz = Settings.System.class;

        public static Member getString = null;

        static {
            try {
                getString = clazz.getDeclaredMethod("getString", ContentResolver.class, String.class);
            } catch (NoSuchMethodException e) {
            }
        }
    }

    public static class TelephonyManager {
        public static final Class<?> clazz = android.telephony.TelephonyManager.class;

        public static Member getDeviceId = null;
        public static Member getDeviceIdWithInt = null;
        public static Member getImei = null;
        public static Member getImeiWithInt = null;
        public static Member getSubscriberId = null;

        static {
            try {
                getDeviceId = clazz.getDeclaredMethod("getDeviceId");
                getDeviceIdWithInt = clazz.getDeclaredMethod("getDeviceId", int.class);
                getImei = clazz.getDeclaredMethod("getImei");
                getImeiWithInt = clazz.getDeclaredMethod("getImei", int.class);
                getSubscriberId = clazz.getDeclaredMethod("getSubscriberId");
            } catch (NoSuchMethodException e) {
            }
        }
    }
}

File: miit_rule_checker/src/main/java/com/loper7/miit_rule_checker/MIITMethods.kt
package com.loper7.miit_rule_checker

import android.annotation.SuppressLint
import android.content.ContentResolver
import android.content.pm.VersionedPackage
import android.location.LocationListener
import android.os.Build
import java.lang.reflect.Member

object MIITMethods {

    @JvmStatic
    fun getDefaultMethods() : MutableList<Member?> {
        return mutableListOf(WifiInfo.getMacAddress ,
            WifiInfo.getIpAddress ,
            LocationManager.getLastKnownLocation ,
            LocationManager.requestLocationUpdates ,
            NetworkInterface.getHardwareAddress ,
            ApplicationPackageManager.getInstalledPackages ,
            ApplicationPackageManager.getInstalledApplications ,
            ApplicationPackageManager.getPackageInfo ,
            ApplicationPackageManager.getInstallerPackageName ,
            PackageManager.getInstalledPackages ,
            PackageManager.getInstalledApplications ,
            PackageManager.getPackageInfo ,
            PackageManager.getPackageInfo1 ,
            PackageManager.getPackageInfo2 ,
            PackageManager.getPackageInfo3 ,
            PackageManager.getInstallerPackageName ,
            Secure.getString ,
            TelephonyManager.getDeviceId ,
            TelephonyManager.getDeviceIdWithInt ,
            TelephonyManager.getImei ,
            TelephonyManager.getImeiWithInt ,
            TelephonyManager.getSubscriberId)
    }
    object WifiInfo {
        private val clazz = android.net.wifi.WifiInfo::class.java
        @JvmStatic
        val getMacAddress = clazz.getDeclaredMethod("getMacAddress")
        @JvmStatic
        val getIpAddress = clazz.getDeclaredMethod("getIpAddress")
    }

    object LocationManager {
        private val clazz = android.location.LocationManager::class.java
        @JvmStatic
        val getLastKnownLocation = clazz.getDeclaredMethod("getLastKnownLocation" , String::class.java)
        @JvmStatic
        val requestLocationUpdates = clazz.getDeclaredMethod("requestLocationUpdates" ,
            String::class.java ,
            Long::class.javaPrimitiveType ,
            Float::class.javaPrimitiveType ,
            LocationListener::class.java)
    }

    object NetworkInterface {
        private val clazz = java.net.NetworkInterface::class.java
        @JvmStatic
        val getHardwareAddress = clazz.getDeclaredMethod("getHardwareAddress")
    }

    object ApplicationPackageManager {
        @SuppressLint("PrivateApi")
        private val clazz : Class<*>? = try {
            Class.forName("android.app.ApplicationPackageManager")
        } catch (e : Exception) {
            null
        }
        @JvmStatic
        val getInstalledPackages = clazz?.getDeclaredMethod("getInstalledPackages" , Int::class.javaPrimitiveType)
        @JvmStatic
        val getInstalledApplications =
            clazz?.getDeclaredMethod("getInstalledApplications" , Int::class.javaPrimitiveType)
        @JvmStatic
        val getInstallerPackageName = clazz?.getDeclaredMethod("getInstallerPackageName" , String::class.java)
        @JvmStatic
        val getPackageInfo =
            clazz?.getDeclaredMethod("getPackageInfo" , String::class.java , Int::class.javaPrimitiveType)
    }

    object PackageManager {
        private val clazz = android.content.pm.PackageManager::class.java
        @JvmStatic
        val getInstalledPackages = clazz.getDeclaredMethod("getInstalledPackages" , Int::class.javaPrimitiveType)
        @JvmStatic
        val getInstalledApplications =
            clazz.getDeclaredMethod("getInstalledApplications" , Int::class.javaPrimitiveType)
        @JvmStatic
        val getInstallerPackageName = clazz.getDeclaredMethod("getInstallerPackageName" , String::class.java)
        @JvmStatic
        val getPackageInfo =
            clazz.getDeclaredMethod("getPackageInfo" , String::class.java , Int::class.javaPrimitiveType)
        @JvmStatic
        val getPackageInfo1 = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            clazz.getDeclaredMethod("getPackageInfo" , String::class.java , android.content.pm.PackageManager.PackageInfoFlags::class.java)
        } else {
            null
        }
        @JvmStatic
       val getPackageInfo2 = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
           clazz.getDeclaredMethod("getPackageInfo" , VersionedPackage::class.java ,Int::class.javaPrimitiveType)
       } else {
           null
       }
        @JvmStatic
        val getPackageInfo3 = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            clazz.getDeclaredMethod("getPackageInfo" , VersionedPackage::class.java, android.content.pm.PackageManager.PackageInfoFlags::class.java)
        } else {
            null
        }
    }

    object Secure {
        private val clazz = android.provider.Settings.Secure::class.java
        @JvmStatic
        val getString = clazz.getDeclaredMethod("getString" , ContentResolver::class.java , String::class.java)
    }

    object TelephonyManager {
        private val clazz = android.telephony.TelephonyManager::class.java
        @JvmStatic
        val getDeviceId = clazz.getDeclaredMethod("getDeviceId")
        @JvmStatic
        val getDeviceIdWithInt = clazz.getDeclaredMethod("getDeviceId" , Int::class.javaPrimitiveType)
        @JvmStatic
        val getImei = clazz.getDeclaredMethod("getImei")
        @JvmStatic
        val getImeiWithInt = clazz.getDeclaredMethod("getImei" , Int::class.javaPrimitiveType)
        @JvmStatic
        val getSubscriberId = clazz.getDeclaredMethod("getSubscriberId")
    }


}